File: js/test/Exchange/test.market.js
'use strict'

const assert = require ('assert');
const testCommonItems = require ('./test.commonItems.js');

function testMarket (exchange, market, method) {
    const method = 'testMarket';
    const format = {
        'id': 'btcusd', // string literal for referencing within an exchange
        'symbol': 'BTC/USD', // uppercase string literal of a pair of currencies
        'base': 'BTC', // unified uppercase string, base currency, 3 or more letters
        'quote': 'USD', // unified uppercase string, quote currency, 3 or more letters
        'taker': exchange.parseNumber ('0.0011'), // taker fee, for example, 0.0011 = 0.11%
        'maker': exchange.parseNumber ('0.0009'), // maker fee, for example, 0.0009 = 0.09%
        'baseId': 'btc', // exchange-specific base currency id
        'quoteId': 'usd', // exchange-specific quote currency id
        'active': true, // boolean, market status
        'type': 'spot',
        'linear': undefined,
        'inverse': undefined,
        'spot': true,
        'swap': false,
        'future': false,
        'option': false,
        'margin': false,
        'contract': false,
        'contractSize': exchange.parseNumber ('0.001'),
        'expiry': 1656057600000,
        'expiryDatetime': '2022-06-24T08:00:00.000Z',
        'optionType': 'put',
        'strike': exchange.parseNumber ('56000'),
        'settle': undefined,
        'settleId': undefined,
        'precision': {
            // todo : handle precision types after another PR is merged
            'price': exchange.parseNumber ('8'), // integer or fraction
            'amount': exchange.parseNumber ('8'), // integer or fraction
            'cost': exchange.parseNumber ('8'), // integer or fraction
        },
        // value limits when placing orders on this market
        'limits': {
            'amount': {
                'min': exchange.parseNumber ('0.01'), // order amount should be > min
                'max': exchange.parseNumber ('1000'), // order amount should be < max
            },
            'price': {
                'min': exchange.parseNumber ('0.01'), // order price should be > min
                'max': exchange.parseNumber ('1000'), // order price should be < max
            },
            // order cost = price * amount
            'cost': {
                'min': exchange.parseNumber ('0.01'), // order cost should be > min
                'max': exchange.parseNumber ('1000'), // order cost should be < max
            },
        },
        'info': {}, // the original unparsed market info from the exchange
    };
    const emptyNotAllowedFor = [ 'id', 'symbol', 'base', 'quote', 'baseId', 'quoteId', 'precision', 'limits', 'type', 'spot', 'margin', 'swap', 'future', 'contract', 'info' ];
    testCommonItems.testStructureKeys (exchange, method, market, format, emptyNotAllowedFor);
    const logText = testCommonItems.logTemplate (exchange, method, market);
    //
    testCommonItems.Gt (exchange, method, market, 'contractSize', '0');
    testCommonItems.Ge (exchange, method, market, 'expiry', '0');
    testCommonItems.Ge (exchange, method, market, 'strike', '0');
    testCommonItems.checkAgainstArray (exchange, method, market, 'optionType', [ 'put', 'call' ]);
    // todo: handle str/num types later
    // assert ((market['taker'] === undefined) || (typeof market['taker'] === 'number'));
    // assert ((market['maker'] === undefined) || (typeof market['maker'] === 'number'));
    if (market['contract']) {
        assert (market['linear'] !== market['inverse'], 'market linear and inverse must not be the same' + logText);
    } else {
        assert ((market['linear'] === undefined) && (market['inverse'] === undefined), 'market linear and inverse must be undefined when "contract" is true' + logText);
    }
    if (market['option']) {
        assert (market['strike'] !== undefined, '"strike" must be defined when "option" is true' + logText);
        assert (market['optionType'] !== undefined, '"optionType" must be defined when "option" is true' + logText);
    }
    const validTypes = [ 'spot', 'margin', 'swap', 'future', 'option' ];
    testCommonItems.checkAgainstArray (exchange, method, market, 'type', validTypes);
    const types = validTypes;
    for (let i = 0; i < types.length; i++) {
        testCommonItems.checkAgainstArray (exchange, method, market, types[i], [ true, false, undefined ]);
    }
    if (market['future']) {
        assert ((market['swap'] === false) && (market['option'] === false), 'market swap and option must be false when "future" is true' + logText);
    } else if (market['swap']) {
        assert ((market['future'] === false) && (market['option'] === false), 'market future and option must be false when "swap" is true' + logText);
    } else if (market['option']) {
        assert ((market['future'] === false) && (market['swap'] === false), 'market future and swap must be false when "option" is true' + logText);
    }
    if (market['linear']) {
        assert (market['inverse'] === false, 'market inverse must be false when "linear" is true' + logText);
    } else if (market['inverse']) {
        assert (market['linear'] === false, 'market linear must be false when "inverse" is true' + logText);
    }
    if (market['future']) {
        assert (market['expiry'] !== undefined, '"expiry" must be defined when "future" is true' + logText);
        assert (market['expiryDatetime'] !== undefined, '"expiryDatetime" must be defined when "future" is true' + logText);
    }
    if (market['expiry'] !== undefined) {
        assert (market['expiryDatetime'] === exchange.iso8601 (market['expiry']), 'expiryDatetime must be equal to expiry in iso8601 format' + logText);
    }
}

module.exports = testMarket;

File: python/ccxt/test/test_market.py
import numbers  # noqa: E402


def test_market(exchange, market, method):
    format = {
        'id': 'btcusd',  # string literal for referencing within an exchange
        'symbol': 'BTC/USD',  # uppercase string literal of a pair of currencies
        'base': 'BTC',  # unified uppercase string, base currency, 3 or more letters
        'quote': 'USD',  # unified uppercase string, quote currency, 3 or more letters
        'taker': 0.0011,  # taker fee, for example, 0.0011 = 0.11%
        'maker': 0.0009,  # maker fee, for example, 0.0009 = 0.09%
        'baseId': 'btc',  # exchange-specific base currency id
        'quoteId': 'usd',  # exchange-specific quote currency id
        'active': True,  # boolean, market status
        'type': 'spot',
        'linear': None,
        'inverse': None,
        'spot': True,
        'swap': False,
        'future': False,
        'option': False,
        'margin': False,
        'contract': False,
        'contractSize': 0.001,
        'expiry': 1656057600000,
        'expiryDatetime': '2022-06-24T08:00:00.000Z',
        'optionType': 'put',
        'strike': 56000,
        'settle': None,
        'settleId': None,
        'precision': {
            'price': 8,  # integer or fraction
            'amount': 8,  # integer or fraction
            'cost': 8,  # integer or fraction
        },
        # value limits when placing orders on self market
        'limits': {
            'amount': {
                'min': 0.01,  # order amount should be > min
                'max': 1000,  # order amount should be < max
            },
            'price': {
                'min': 0.01,  # order price should be > min
                'max': 1000,  # order price should be < max
            },
            # order cost = price * amount
            'cost': {
                'min': 0.01,  # order cost should be > min
                'max': 1000,  # order cost should be < max
            },
        },
        'info': {},  # the original unparsed market info from the exchange
    }
    keys = list(format.keys())
    for i in range(0, len(keys)):
        key = keys[i]
        keyPresent = (key in market)
        assert keyPresent, key + ' missing ' + exchange.json(market)

    keys = [
        'id',
        'symbol',
        'baseId',
        'quoteId',
        'base',
        'quote',
        'precision',
        'limits',
    ]
    for i in range(0, len(keys)):
        key = keys[i]
        assert market[key] is not None, key + ' None ' + exchange.json(market)

    assert(market['taker'] is None) or (isinstance(market['taker'], numbers.Real))
    assert(market['maker'] is None) or (isinstance(market['maker'], numbers.Real))
    if market['contract']:
        assert market['linear'] != market['inverse']
    else:
        assert(market['linear'] is None) and (market['inverse'] is None)

    if market['option']:
        assert market['strike'] is not None
        assert market['optionType'] is not None

    validTypes = {
        'spot': True,
        'margin': True,
        'swap': True,
        'future': True,
        'option': True,
    }
    type = market['type']
    #
    # binance has type = 'delivery'
    # https://github.com/ccxt/ccxt/issues/11121
    #
    # assert type in validTypes
    #
    types = list(validTypes.keys())
    for i in range(0, len(types)):
        entry = types[i]
        if entry in market:
            value = market[entry]
            assert(value is None) or value or not value

    #
    # todo fix binance
    #
    # if market['future']:
    #     assert(market['swap'] == False) and (market['option'] == False)
    # elif market['swap']:
    #     assert(market['future'] == False) and (market['option'] == False)
    # elif market['option']:
    #     assert(market['future'] == False) and (market['swap'] == False)
    # }
    # if market['linear']:
    #     assert market['inverse'] == False
    # elif market['inverse']:
    #     assert market['linear'] == False
    # }
    # if market['future']:
    #     assert market['expiry'] is not None
    #     assert market['expiryDatetime'] is not None
    # }
